pennylane_lightning/core/src/simulators/lightning_gpu/observables/ObservablesGPU.hpp

```diff
@@ -0,0 +1,287 @@
+     */
+    template <typename T1, typename T2>
+    explicit Hamiltonian(T1 &&coeffs, T2 &&obs) : BaseType{coeffs, obs} {}
+
+    /**
+     * @brief Convenient wrapper for the constructor as the constructor does not
+     * convert the std::shared_ptr with a derived class correctly.
+     *
+     * This function is useful as std::make_shared does not handle
+     * brace-enclosed initializer list correctly.
+     *
+     * @param coeffs Argument to construct coefficients
+     * @param obs Argument to construct terms
+     */
+    static auto
+    create(std::initializer_list<PrecisionT> coeffs,
+           std::initializer_list<std::shared_ptr<Observable<StateVectorT>>> obs)
+        -> std::shared_ptr<Hamiltonian<StateVectorT>> {
+        return std::shared_ptr<Hamiltonian<StateVectorT>>(
+            new Hamiltonian<StateVectorT>{std::move(coeffs), std::move(obs)});
+    }
+
+    // to work with
+    void applyInPlace(StateVectorT &sv) const override {
+        using CFP_t = typename StateVectorT::CFP_t;
+        DataBuffer<CFP_t, int> buffer(sv.getDataBuffer().getLength(),
+                                      sv.getDataBuffer().getDevTag());
+        buffer.zeroInit();
+
+        for (size_t term_idx = 0; term_idx < this->coeffs_.size(); term_idx++) {
+            StateVectorT tmp(sv);
+            this->obs_[term_idx]->applyInPlace(tmp);
+            scaleAndAddC_CUDA(
+                std::complex<PrecisionT>{this->coeffs_[term_idx], 0.0},
+                tmp.getData(), buffer.getData(), tmp.getLength(),
+                tmp.getDataBuffer().getDevTag().getDeviceID(),
+                tmp.getDataBuffer().getDevTag().getStreamID(),
+                tmp.getCublasCaller());
+        }
+        sv.CopyGpuDataToGpuIn(buffer.getData(), buffer.getLength());
+    }
+};
+
+/**
+ * @brief Sparse representation of HamiltonianGPU<StateVectorT>
+ *
+ */
+template <class StateVectorT>
+class SparseHamiltonian final : public SparseHamiltonianBase<StateVectorT> {
+  public:
+    using PrecisionT = typename StateVectorT::PrecisionT;
+    using ComplexT = typename StateVectorT::ComplexT;
+    // cuSparse required index type
+    using IdxT =
+        typename std::conditional<std::is_same<PrecisionT, float>::value,
+                                  int32_t, int64_t>::type;
+
+  private:
+    using BaseType = SparseHamiltonianBase<StateVectorT>;
+
+  public:
+    /**
+     * @brief Create a SparseHamiltonian from data, indices and offsets in CSR
+     * format.
+     *
+     * @param data Arguments to construct data
+     * @param indices Arguments to construct indices
+     * @param offsets Arguments to construct offsets
+     * @param wires Arguments to construct wires
+     */
+    template <typename T1, typename T2, typename T3 = T2, typename T4>
+    explicit SparseHamiltonian(T1 &&data, T2 &&indices, T3 &&offsets,
+                               T4 &&wires)
+        : BaseType{data, indices, offsets, wires} {}
```

Comment:
Do we need to worry about the data here using the base-type `std::complex<T>` type and the vector of data?

Reply:
No need now since it's removed.